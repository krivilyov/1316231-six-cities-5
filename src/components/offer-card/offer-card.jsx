import React from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {offerPropType} from "../../prop-types";
import {OfferCardType} from "../../const";
import OfferCardBookmark from "../offer-card-bookmark/offer-card-bookmark";
import {connect} from "react-redux";
import {fetchBookmarkOffers, updateOfferBookmarkStatus} from "../../store/api-actions";
import {setOverOfferId} from "../../store/action";

const OfferCard = (props) => {
  const {offer, currentCardType, overOfferId, onChangeOfferId} = props;
  const {id, isPremium, smallImg, price, rating, title, type} = offer;
  const isFavoriteType = currentCardType === OfferCardType.FAVORITE;

  const getCardClass = {
    index: `cities__place-card`,
    related: `near-places__card`,
    favorite: `favorites__card`,
  };

  const getWrapperClass = {
    index: `cities__image-wrapper`,
    related: `near-places__image-wrapper`,
    favorite: `favorites__image-wrapper`,
  };

  const onSetNewId = () => {
    if (offer.id !== overOfferId) {
      onChangeOfferId(offer.id);
    }
  };

  const checkIsFavorite = () => isFavoriteType ? `favorites__card-info` : ``;

  return (
    <article className={`${getCardClass[currentCardType]} place-card`}
      onMouseEnter={currentCardType === OfferCardType.INDEX ? () => {
        onSetNewId();
      } : undefined}
      onMouseLeave={currentCardType === OfferCardType.INDEX ? () => {
        onChangeOfferId(``);
      } : undefined}>
      {isPremium && (
        <div className="place-card__mark">
          <span>Premium</span>
        </div>)}
      <div className={`${getWrapperClass[currentCardType]} place-card__image-wrapper`}>
        <Link to={`/offer/${id}`} className={`place-card__img-link`}
          onClick={currentCardType !== OfferCardType.INDEX ? onSetNewId : undefined}
        >
          <img className="place-card__image"
            src={smallImg}
            width={isFavoriteType ? `150` : `260`}
            height={isFavoriteType ? `110` : `200`}
            alt="Place image"/>
        </Link>
      </div>
      <div className={`place-card__info ${checkIsFavorite()}`}>
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>

          <OfferCardBookmark
            offerId={offer.id}
            offerBookmarkStatus={offer.isBookmark}
          />

        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `${rating * 20}%`}}></span>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <Link to={`/offer/${id}`}
            onClick={currentCardType !== OfferCardType.INDEX ? onSetNewId : undefined}
          >{title}</Link>
        </h2>
        <p className="place-card__type">{type}</p>
      </div>
    </article>
  );
};

OfferCard.propTypes = {
  offer: offerPropType,
  currentOffer: offerPropType,
  currentCardType: PropTypes.string.isRequired,
  onChangeOfferId: PropTypes.func.isRequired,
  overOfferId: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  overOfferId: state.COMMON.overOfferId
});

const mapDispatchToProps = (dispatch) => ({
  onChangeOfferId(offerId) {
    dispatch(setOverOfferId(offerId));
  },
  onChangeBookmark(offerId, bookmarkStatus) {
    dispatch(updateOfferBookmarkStatus(offerId, bookmarkStatus));
  },
  onChangeBookmarkOffers() {
    dispatch(fetchBookmarkOffers());
  },
});

export {OfferCard};
export default connect(mapStateToProps, mapDispatchToProps)(OfferCard);
