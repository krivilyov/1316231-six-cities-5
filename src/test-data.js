export const testInitialState = {
  COMMON: {
    activeCity: `Amsterdam`,
    offers: [],
    reviews: [],
    sortingType: `popular`,
    mouseOverOfferId: null,
    changedBookmarkOffer: {},
    relatedOffers: [],
    bookmarkOffers: [],
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    authInfo: {},
  },
};
