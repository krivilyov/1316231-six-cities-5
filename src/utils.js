export function extend(obj1, obj2) {
  return Object.assign({}, obj1, obj2);
}
