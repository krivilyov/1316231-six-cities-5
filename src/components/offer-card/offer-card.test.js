import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import configureMockStore from "redux-mock-store";
import {OfferCard} from "./offer-card";
import {testInitialState} from "../../test-data";
import {formatUpperCaseFirst} from "../../utils";

const mockStore = configureMockStore();
const store = mockStore(testInitialState);

const offer = {
  id: `7`,
  img: [`url1`, `url2`, `url3`, `url4`, `url5`, `url6`],
  smallImg: `small-url`,
  title: `title`,
  description: [`description`],
  isPremium: false,
  type: formatUpperCaseFirst(`type1`),
  rating: 4.8,
  bedroom: 3,
  visitor: 4,
  price: 256,
  option: [`option1`, `option2`, `option3`, `option4`],
  isBookmark: false,
  city: `city1`,
  cityCoordinates: [48.85661, 2.351499],
  cityZoom: 13,
  host: `hostName`,
  hostTop: false,
  avatar: `avatar_url`,
  coordinates: [48.877610000000004, 2.333499],
};
const noop = () => {};

it(`OfferCard should render correctly`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <MemoryRouter>
            <OfferCard
              offer={offer}
              changedBookmarkOffer={offer}
              currentCardType={`main`}
              overOfferId={`1`}
              onMouseOverOffer={noop}
              onChangeOfferId={noop}
              onChangeBookmarkOffers={noop}
              onChangeBookmark={noop}
            />
          </MemoryRouter>
        </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
