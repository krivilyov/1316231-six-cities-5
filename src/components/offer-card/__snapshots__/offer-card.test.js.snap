// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferCard should render correctly 1`] = `
<article
  className="undefined place-card"
  onMouseOver={[Function]}
>
  <div
    className="undefined place-card__image-wrapper"
  >
    <a
      href="/offer/7"
      onClick={[Function]}
    >
      <img
        alt="Place image"
        className="place-card__image"
        height="200"
        src="small-url"
        width="260"
      />
    </a>
  </div>
  <div
    className="place-card__info "
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          256
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <a
        className="place-card__bookmark-button
         button"
        href="/login"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          To
           bookmarks
        </span>
      </a>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "96%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="/offer/7"
        onClick={[Function]}
      >
        title
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      Type1
    </p>
  </div>
</article>
`;
